### PR TITLE
Nix flatten solidity source

### PR DESCRIPTION
**Pull request by icetan**
- Add `flatten` argument to `solidityPackage` to pass source through `hevm flatten`
- Fix to allow non-`solidity` `pragmas` being passed through `hevm flatten`
- Rename top scoped contracts and structs so not to cause naming conflicts on flatten

### CODE REVIEW

File: nix/solidity-package.nix [hunk 1]
```diff
@@ -19,12 +19,14 @@ in
...
     , test-hevm ? pkgs.dapp2.test-hevm
+    , hevm ? pkgs.hevm
     , solcFlags ? ""
+    , flatten ? false
     , ...
     } @ attrs:
       pkgs.stdenv.mkDerivation (rec {
         inherit doCheck;
-        buildInputs = [ test-hevm solc ];
+        buildInputs = [ test-hevm solc ] ++ (if flatten then [hevm] else []);
```
Comment (asymmetric) [inline nix/solidity-package.nix]:
```suggestion
        buildInputs = [ test-hevm solc ] ++ pkgs.lib.optional flatten hevm;
```

**Comment by MrChico** [comment]
What's the status of this? @icetan

**Comment by icetan** [comment]
Working but needs a cleanup before proper review.

**Comment by brianmcmichael** [comment]
+1 this update now handles most of the usecases for the Maker contracts, including support for `pragma experimental abiencoderv2`

There is still an issue with some contracts where an imported contract includes contract name conflicts with other imported contracts ([dss-deploy](https://github.com/makerdao/dss-deploy) is the offender I'm talking about), but none of the other flatteners on the market can handle this properly either. The version on etherscan now was manually flattened.

The benefits here are worth adding in now and then tackling the issues we're facing in dss-deploy later, most users aren't going to be verifying contracts that complex right now anyways.

**Comment by asymmetric** [comment]
@icetan what are the main benefits of this PR?

**Pull request merged**